_posts/2016-06-29-concurrency-in-go.md

```diff
@@ -0,0 +1,69 @@
+Web services today is largely dependent upon I/O. Either from disk, database or an external service. Running these operations sequentially and waiting for them to finish will result in a slow and underperforming system. Most modern web frameworks solves the basic issues for you. That is, without setup it handles each http request concurrently. But if you need to do something out of the ordinary, like calling a few external services and combined the results you are mostly on your own.
+
+The two most common models for concurrency that I’ve used is shared-memory model using Threads like in Java. Or callbacks used in asynchronously languages like in Node.js. I believe that both approaches can be insanely powerful when done right. However, that they’re also insanely hard to get right. Shared-memory model sharing state/messages through memory using locks and is error-prone to say the least. And asynchronous programming is, at least in my experience, a hard programming paradigm reason about and especially to master.
+
+## Concurrency in Go
+Go solves concurrency in a different manner. It’s similar to Threads but instead of sharing messages through memory, go shares memory through messages. Go uses goroutines to achieve concurrency and channels for passing data between them. We will dig into these two concepts a bit further.
+
+### Goroutines
+
+Goroutines is a simple abstraction for running things (functions) concurrently.  This is achieved by prepending ``go`` before a function call. E.g.
+
+<script src="https://gist.github.com/gronnbeck/9c363b773e7bb43b4e58cea67ba8cb89.js"></script>
+
+A good example of the concept can be found [here](https://tour.golang.org/concurrency/1)
+
+### Channels
+
+Channels is the construct for passing data between routines in Go. A channel blocks both on the sending and receiving side until both are ready. Meaning a channel can be used for both synchronising goroutines and passing data between them. Below we see a simple example of how to use channels in Go. The basic idea is that data flows the same directions as the arrow.
+
+<script src="https://gist.github.com/gronnbeck/9c363b773e7bb43b4e58cea67ba8cb89.js"></script>
+
+In the example below we see how channels and goroutines can be used to create a function utilising concurrency that is easy to understand and reason about.
+
+## Example: using goroutines and channels
+
+First let’s assume we want to create service that asks three external services and return them. Let’s call these three services Facebook, Twitter and Github. For simplicity, we a fake communication with each of these services, such that the result of each service can be found at ``https://localhost/{facebook, twitter, github}``, respectively.
+
+The behaviour for ``GetAllTheThings`` is to fetch data from all services defined and combined them into a list. Let’s start with a naive approach.
+
+<script src="https://gist.github.com/gronnbeck/6caa0bac97e217e9542d41ab35398da4.js"></script>
```

Comment:
Unidiomatic to not handle err in go. Handle them to not wake some Internet beast?